core/offchain/impl/http_request.cpp

```diff
@@ -0,0 +1,433 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "offchain/impl/http_request.hpp"
+
+#include <algorithm>
+#include <boost/asio/connect.hpp>
+#include <boost/asio/ip/tcp.hpp>
+#include <boost/asio/ssl/context.hpp>
+#include <boost/asio/ssl/error.hpp>
+#include <boost/asio/ssl/stream.hpp>
+#include <boost/beast/http.hpp>
+#include <string>
+#include <string_view>
+#include <thread>
+
+namespace kagome::offchain {
+
+  HttpRequest::HttpRequest(RequestId id)
+      : id_(id),
+        resolver_(io_context_),
+        ssl_ctx_(boost::asio::ssl::context::sslv23),
+        deadline_timer_(io_context_),
+        log_(log::createLogger("HttpRequest#" + std::to_string(id_),
+                               "offchain")) {
+    ssl_ctx_.set_default_verify_paths();
+    ssl_ctx_.set_verify_mode(boost::asio::ssl::verify_peer);
+    ssl_ctx_.set_verify_callback(
+        [log = log_, wp = weak_from_this()](
+            bool preverified, boost::asio::ssl::verify_context &ctx) {
+          // We will simply print the certificate's subject name here
+          char subject_name[256];
+          X509 *cert = X509_STORE_CTX_get_current_cert(ctx.native_handle());
+          X509_NAME_oneline(X509_get_subject_name(cert), subject_name, 256);
+          SL_WARN(log,
+                  "Verifying [{}] was {}",
+                  subject_name,
+                  preverified ? "Successful" : "Failed");
+
+          return preverified;
+        });
+  }
+
+  bool HttpRequest::init(HttpMethod method,
```

Comment:
Why not return an outcome::result and leave its processing to the calling side? I think this might be beneficial in the future, who knows how this class will be used.

Reply:
Not sure if upstream should handle error. It is just mediator between runtime and api, and has binary logic of result.